source/MRMesh/MRMeshLoad3mf.cpp

```diff
@@ -0,0 +1,331 @@
+            return unexpected( std::string( "3DF model vertex node does not have 'x' attribute" ) );
+        if ( tinyxml2::XML_SUCCESS != vertexNode->QueryFloatAttribute( "y", &p.y ) )
+            return unexpected( std::string( "3DF model vertex node does not have 'y' attribute" ) );
+        if ( tinyxml2::XML_SUCCESS != vertexNode->QueryFloatAttribute( "z", &p.z ) )
+            return unexpected( std::string( "3DF model vertex node does not have 'z' attribute" ) );
+        vertexCoordinates.push_back( p );
+    }
+
+    if ( !reportProgress( callback, 0.25f ) )
+        return unexpected( std::string( "Loading canceled" ) );
+
+    auto trianglesNode = meshNode->FirstChildElement( "triangles" );
+    if ( !trianglesNode )
+        return unexpected( std::string( "3DF model 'triangles' node not found" ) );
+
+    Triangulation tris;
+    for ( auto triangleNode = trianglesNode->FirstChildElement( "triangle" ); triangleNode; triangleNode = triangleNode->NextSiblingElement( "triangle" ) )
+    {
+        int vs[3];
+        if ( tinyxml2::XML_SUCCESS != triangleNode->QueryIntAttribute( "v1", &vs[0] ) )
+            return unexpected( std::string( "3DF model triangle node does not have 'v1' attribute" ) );
+        if ( tinyxml2::XML_SUCCESS != triangleNode->QueryIntAttribute( "v2", &vs[1] ) )
+            return unexpected( std::string( "3DF model triangle node does not have 'v2' attribute" ) );
+        if ( tinyxml2::XML_SUCCESS != triangleNode->QueryIntAttribute( "v3", &vs[2] ) )
+            return unexpected( std::string( "3DF model triangle node does not have 'v3' attribute" ) );
+        tris.push_back( { VertId( vs[0] ), VertId( vs[1] ), VertId( vs[2] ) } );
+    }
+
+    if ( !reportProgress( callback, 0.5f ) )
+        return unexpected( std::string( "Loading canceled" ) );
+
+    FaceBitSet skippedFaces;
+    std::vector<MeshBuilder::VertDuplication> dups;
+    std::vector<MeshBuilder::VertDuplication>* dupsPtr = nullptr;
+    if ( loader.duplicatedVertexCount )
+        dupsPtr = &dups;
+    MeshBuilder::BuildSettings buildSettings;
+    if ( loader.skippedFaceCount )
+    {
+        skippedFaces = FaceBitSet( tris.size() );
+        skippedFaces.set();
+        buildSettings.region = &skippedFaces;
+    }
+    MR::Mesh mesh = Mesh::fromTrianglesDuplicatingNonManifoldVertices( std::move( vertexCoordinates ), tris, dupsPtr, buildSettings );
+    if ( loader.duplicatedVertexCount )
+        *loader.duplicatedVertexCount += int( dups.size() );
+    if ( loader.skippedFaceCount )
+        *loader.skippedFaceCount += int( skippedFaces.count() );
+
+    if ( !reportProgress( callback, 0.75f ) )
+        return unexpected( std::string( "Loading canceled" ) );
+
+    return std::make_shared<Mesh>( std::move( mesh ) );
+};
+
+// Parse multi-component object form <components> or <build> element
+Expected<std::shared_ptr<Mesh>, std::string> loadComponents(
```

Comment:
Can we return just `Mesh` and not `std::shared_ptr<Mesh>`?

Reply:
And in other places

Reply:
> Can we return just `Mesh` and not `std::shared_ptr<Mesh>`?
When a mesh is being loaded in a scene build, it is both used in the scene and put into a resources cache for further usage. Shared pointers should eliminate extra copies when there is no transformation.
Probably not worth it, using just `Mesh` would make code simpler.